### ISSUE TITLE
How to generate mocap.npz?

### ISSUE THREAD

**Issue by Minotaur-CN** [electronicarts/character-motion-vaes#1]
Hi,
How to generage mocap.npz, which seems not easy to me. 
Can u give a clue how to generate mocap.npz from public mocap dataset?

The train_mvae.py script assumes the mocap data to be at environments/mocap.npz. The original training data is not included in this repo; but can be easily extracted from other public datasets.

Thanks very much!

BEST

**Comment by wangshub** [comment]
+1, could you provide some details and format description about the mocap data? Thanks!

@fabiozinno @belinghy

**Comment by OOF-dura** [comment]
+1, could you provide some details and format description about the mocap data? Thanks!

**Comment by wangshub** [comment]
@Minotaur-CN @OOF-dura 

the raw data format is not complex, just read the code below

- https://github.com/electronicarts/character-motion-vaes/blob/main/vae_motion/train_mvae.py#L140
- https://github.com/electronicarts/character-motion-vaes/blob/main/vae_motion/train_mvae.py#L141

**Comment by belinghy** [comment]
Below is some information about the data format.  We also note the length of each mocap sequence (as mentioned above at L141).  This is so we don't sample invalid transitions for training.  If the mocap clip is one long continuous sequence, then there no reason to do this. 

```
    0-3 : root delta x, delta y, delta facing
   3-69 : joint coordinates (22 * 3 = 66)
 69-135 : joint velocities in Cartesian coordinate in previous root frame (22 * 3 = 66)
135-267 : 6D joint orientations, i.e. first two columns of rotation matrix (22 * 6 = 132)
```

For extracting training data from mocap datasets, I think [fairmotion](https://github.com/facebookresearch/fairmotion) might be helpful.  Based on the examples I have seen, though I haven't tested, should be something like below.  Root deltas need some more processing; essentially find the displacement vector and rotate by the current facing direction of the character.  Same thing for positions and velocities, they should be projected to the character space to make learning easier.

```
from fairmotion.data import bvh

motion = bvh.load(BVH_FILENAME)

positions = motion.positions(local=False)  # (frames, joints, 3)
velocities = positions[1:] - positions[:-1]
orientations = motion.rotations(local=False)[..., :, :2].reshape(-1, 22, 6)
```

**Comment by Realdr4g0n** [comment]
@belinghy can you tell me more about how to get root deltas?

I think a sample, formula or code would be better

**Comment by ameliacode** [comment]
I may have misunderstood the whole process but since there aren't any sample of mocap.npz, I assume that mocap.npz should be like this 

mocap.npz
"data": list of 267 float numbers (the first info about root delta is included in paper "pose representation")
"end_indices": end_frame?( length of each mocap sequence )

**Comment by ameliacode** [comment]
I may have misunderstood the whole process but since there aren't any sample of mocap.npz, I assume that mocap.npz should be like this:

mocap.npz
"data": list of 267 float numbers (the first info about root delta is included in paper "pose representation")  * frame
"end_indices": 267? ( length of each mocap sequence ) 

It seems mocap data has to include only 22 joints, so, extracting from other public datasets may not work as bvh files or other mocap data out there may have different number of joints.
Even if discarding irrelevant joints from the data, joint index order is another issue as you can see in mocap_env.py :(

Therefore.. I think there are two ways to solve:
- modify joint relevant parts in this project (mocap_env.py , etc.)
- or discard joints and keep joint index sequence refer to pose0.npy ( not sure about this ) 
 
I wasn't able to find what mocap database this project had used. and it wasn't in the paper..:(

**Comment by belinghy** [comment]
Your understanding of the format is correct, except `end_indices` marks the end of mocap clips.  It depends on the number of mocap clips you have, so not necessarily 267.  For example, if there are two clips of lengths 10 and 15, then `end_indices = np.cumsum([10, 15]) - 1 = [9, 24]`.

As you've noted, mocap_env.py could definitely refactored.  I think the only things to change if you are using different input format are [these lines](https://github.com/electronicarts/character-motion-vaes/blob/d8ff62f3a9f7fd330ebd797df3e7f8f7ec7c3816/environments/mocap_envs.py#L86-L94) and [these lines](https://github.com/electronicarts/character-motion-vaes/blob/d8ff62f3a9f7fd330ebd797df3e7f8f7ec7c3816/environments/mocap_envs.py#L136-L137). The second reference is only if `0-3 : root delta x, delta y, delta facing`. Am I missing anything else?

**Comment by ameliacode** [comment]
So.. as it mentioned above, if I get this right, end_indices might contain one integer value if an input clip is a long continuous sequence,
However, I still don't get what "length" is in this case. Is it a frame number? or is there other unit used?

**Comment by belinghy** [comment]
Yes, it's a frame number. end_indices contains one integer value if there is exactly one input clip is a long continuous sequence.

**Comment by edentliang** [comment]
Hi, I have some confusion about 
` 135-267 : 6D joint orientations, i.e. first two columns of rotation matrix (22 * 6 = 132)`
`orientations = motion.rotations(local=False)[..., :, :2].reshape(-1, 22, 6)` 
in your case, is 'z-axis' the world up vector, and 6D joint orientations are the orientations of other two directions?

Furthermore, can you provide some examples for `0-3 : root delta x, delta y, delta facing`? I am a  bit confused about the definition of these variables.

Thank you

**Comment by ameliacode** [comment]
Maybe this will help: https://arxiv.org/pdf/2103.14274.pdf : see pose representation for root information.

I think the paper and code is slightly different in terms of what up-vector they have used.
Overall, root delta position has to include two values of root position projected on the ground, root facing direction also,
and joint orientations have to include a form of rotation matrix of relative forward and upward vector.

**Comment by Gabriel-Bercaru** [comment]
Hello,

@belinghy , when reshaping the rotation components, as returned by fairmotion:
`orientations = motion.rotations(local=False)[..., :, :2].reshape(-1, 22, 6)`, how would the vector components be distributed?
Considering the first joint in the first frame and the 6 on the third dimension contains 3 components for the first 2 columns of the rotation matrix, would they be laid out like in the first version below or as in the second one?

version 1:
```
orientations[0, 0, 0] = comp1_x
orientations[0, 0, 1] = comp1_y
orientations[0, 0, 2] = comp1_z
orientations[0, 0, 3] = comp2_x
orientations[0, 0, 4] = comp2_y
orientations[0, 0, 5] = comp2_z
```

version 2:
```
orientations[0, 0, 0] = comp1_x
orientations[0, 0, 1] = comp2_x
orientations[0, 0, 2] = comp1_y
orientations[0, 0, 3] = comp2_y
orientations[0, 0, 4] = comp1_z
orientations[0, 0, 5] = comp2_z
```

**Comment by belinghy** [comment]
Hi @Gabriel-Bercaru, I'm not sure what is fairmotion's convention.  Are you rendering the character using joint orientations?  If not, for the purpose of neural network input, the order shouldn't matter.

**Comment by Gabriel-Bercaru** [comment]
Hello, indeed for the input training data, it doesn't really matter, but I was trying to render a mesh over a trained model.

As far as I have seen, rigging makes use of the joint orientations and in order to get them I should convert those 6D orientation vectors to either Euler rotations or quaternions

**Comment by belinghy** [comment]
The way it's indexed, e.g., `[..., :, :2]`, should correspond to version 2.

**Comment by ameliacode** [comment]
Hi again. I finally get to know to preprocess npz ! but.. I still having trouble making pose0.npy

As I am working on 31 joint bvh data,mine gets 375 (data.shape[1]) unlike given pose0.npy in this repository, 
So, I thought pose0.npy, in my case, shape should be [1,375].

However, when I looked up to this part:
```
 def reset_initial_frames(self, frame_index=None):
        # Make sure condition_range doesn't blow up
        self.start_indices.random_(
            0, self.mocap_data.shape[0] - self.num_condition_frames + 1
        )
```
It seems I have to give all starting point of all motion clips overall. (ex. 100 motion clips = [100, 375] ??)
I am still not sure about this part. Any Ideas?

P.S. When I train customed pose0.npy with my own mocap data, result shows exactly like this:
https://github.com/electronicarts/character-motion-vaes/issues/7

**Comment by jeffreyking23** [comment]
> 



> Maybe this will help: https://arxiv.org/pdf/2103.14274.pdf : see pose representation for root information.
> 
> I think the paper and code is slightly different in terms of what up-vector they have used. Overall, root delta position has to include two values of root position projected on the ground, root facing direction also, and joint orientations have to include a form of rotation matrix of relative forward and upward vector.

Hi,amelia,I have some trouble in processing LAFAN1 dataset,can you review my code about processing the dataset? 

for index,bf in enumerate(bvh_files):
    motion = bvh.load(bf)
    positions = motion.positions(local=False)  # (frames, joints, 3)
    frame_number = np.shape(positions)[0]
    rotations = motion.rotations(local=False)
    global_orientations = rotations[:, 0, :]
    axis_angle = rotation_matrix_to_angle_axis(torch.from_numpy(global_orientations).to(device=device))
    axis_angle = axis_angle.cpu().numpy()
    root_positions = np.reshape(positions[:, 0, :], [frame_number, 3])
    root_facings = np.reshape(axis_angle[:,2],[frame_number,1])
    root_delta = np.concatenate([root_positions[:,:2],root_facings],axis=1)
    root_delta = np.reshape(root_delta,[frame_number,1,3])
    positions2 = positions[:,:,:2] - root_delta[:,:,:2]
    new_positions = np.concatenate([positions2,np.expand_dims(positions[:,:,2], axis=2)],axis=2)
    velocities = new_positions[1:] - new_positions[:-1]
    axis_angle[:,2] = 0
    new_global_orientations = batch_rodrigues(torch.from_numpy(axis_angle).to(device=device))
    new_global_orientations = new_global_orientations.cpu().numpy()
    new_global_orientations = np.reshape(new_global_orientations,[frame_number,3,3])
    rotations[:, 0, :] = new_global_orientations
    new_orientations = rotations[..., :, :2].reshape(-1, 22, 6)
    new_orientations = np.reshape(new_orientations[1:,:,:],[frame_number-1,132])
    root_delta = np.reshape(root_delta[1:,:,:],[frame_number-1,3])
    new_positions = np.reshape(new_positions[1:, :, :], [frame_number - 1, 66])
    velocities = np.reshape(velocities, [frame_number - 1, 66])
    if index == 0:
        data = np.concatenate([root_delta,new_positions,velocities,new_orientations],axis=1)
        previous = frame_number-1
    else:
        data_curr = np.concatenate([root_delta,new_positions,velocities,new_orientations],axis=1)
        data = np.concatenate([data,data_curr], axis=0)
        previous = previous + frame_number - 1
    end_indices.append(previous)
